### PR TITLE
[feat] masked cropping function

### PR DESCRIPTION
**Pull request by shijianjian**
### Description

Here is another idea of performing cropping, to crop on a mask then apply on images.

Pros:
- The user does not need to know the interpolation methods (e.g. bilinear or nearest).
- The output will not be modified with different parameter settings (e.g. interpolation, align_corners).
- Seems to have more meaningful gradients (treat all pixels equally) than cropping on images.

Cons:
- In terms of meaningful gradients, I am not sure if it is right.

If this is adopted, I would propose to update the current ```crop_by_boxes``` to ```sampling_by_boxes``` since it is essentially a sampling than cropping.

### Status
**Ready/Work in progress/Hold**

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or new feature that would cause existing functionality to change)
- [ ] New tests added to cover the changes
- [ ] Docstrings/Documentation updated

### PR Implementer
This is a small checklist for the implementation details of this PR.

If there are any questions regarding code style or other conventions check out our 
[summary](https://github.com/kornia/kornia/blob/master/CONTRIBUTING.rst).

- [ ] Did you discuss the functionality or any breaking changes before ?
- [ ] **Pass all tests**: did you test in local ? `make test`
- [ ] Unittests: did you add tests for your new functionality ?
- [ ] Documentations: did you build documentation ? `make build-docs`
- [ ] Implementation: is your code well commented and follow conventions ? `make lint`
- [ ] Docstrings & Typing: has your code documentation and typing ? `make mypy`
- [ ] Update notebooks & documentation if necessary

### KorniaTeam
<details>
  <summary>KorniaTeam workflow</summary>
  
  - [ ] Assign correct label
  - [ ] Assign PR to a reviewer
  - [ ] Does this PR close an Issue? (add `closes #IssueNumber` at the bottom if 
        not already in description)

</details>

### Reviewer
<details>
  <summary>Reviewer workflow</summary>

  - [ ] Do all tests pass? (Unittests, Typing, Linting, Documentation, Environment)
  - [ ] Does the implementation follow `kornia` design conventions?
  - [ ] Is the documentation complete enough ?
  - [ ] Are the tests covering simple and corner cases ?
 
</details>

### CODE REVIEW

**Review by edgarriba — COMMENTED** [review]
I see few things here - you propose 3 different contributions here that I'll split in separate PR.

1. the bbox generation (we need further discussion for this API and where to place it (kornia-core)
2. Adapt augmentation functional with 1 (kornia-augmentation)
3. Implement a utility function (also needs quick discussion about the API) (kornia-core)

I would suggest if the PR is not ready to review to mark as Draft.

To proceed with the new feature I'd go step by step and first open a PR with 1. I believe that this can be an opportunity to create `kornia.geometry.bbox` and place those things there. I'll leave some comments about the API

File: kornia/geometry/transform/crop/crop2d_contrib.py [hunk 1]
```diff
@@ -0,0 +1,182 @@
+from typing import Optional
+
+import kornia
+import torch
+from .crop2d import crop_by_boxes as sampling_by_boxes, validate_bboxes
+
+
+def generate_dst_bbox_from_bbox(boxes: torch.Tensor):
```
Comment (edgarriba) [inline kornia/geometry/transform/crop/crop2d_contrib.py:8]: As I mentioned I think this can be one of the first functions of `kornia.geometry.bbox`.

In this particular case I'd suggest to rename `generate_bbox_from_points` since from my understanding this function takes as input four 2D points and creates a bounding box representation [x, y, h, w]. Am I right ? Besides this needs some asserts and tests.

File: kornia/geometry/transform/crop/crop2d_contrib.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+    points_dst: torch.Tensor = torch.stack([
+        torch.stack([zeros, zeros], dim=-1),
+        torch.stack([dst_w, zeros], dim=-1),
+        torch.stack([dst_w, dst_h], dim=-1),
+        torch.stack([zeros, dst_h], dim=-1)
+    ], dim=1)
+    return points_dst
+
+
+def bbox_to_mask(boxes: torch.Tensor, mask: torch.Tensor) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/transform/crop/crop2d_contrib.py:55]: What the use case of this function ? from the description I don't understand the functionality and the purpose.

File: kornia/geometry/transform/crop/crop2d_contrib.py [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+            _roi_padded = _mask * 0.
+
+        padded.append(_roi_padded)
+
+    return torch.cat(padded)
+
+
+def crop_by_boxes(
+    tensor: torch.Tensor, src_box: torch.Tensor, src_mask: Optional[torch.Tensor] = None
+) -> torch.Tensor:
```
Comment (edgarriba) [inline kornia/geometry/transform/crop/crop2d_contrib.py:119]: is this cropping without using bilinear interpolation right ?

**Review by edgarriba — COMMENTED** [review]
few specific comments on the bbox thing

**Comment by shijianjian** [comment]
@edgarriba  This PR is actually proposing a new way of cropping that based on masks as described in #832.

In our previous implementation, the cropping is actually done by perspective transform, which is doable but the gradients is essentially w.r.t the pixel values and the interpolation methods would break the original pixel values. Also, the random_erasing is not really differentiable.

Since that the nearest interpolation will always output 0 grads, I am proposing to use mask instead, to apply bilinear operation on a mask than the original image.

For random erasing, I propose to:
1. convert target box to mask.
2. ```1 - mask``` to select the correct area.
3. ```input * (1 - mask)``` to apply random erasing.
4. You can see the gradients output in the written doctest. The gradients are essentially equal everywhere in the targetted area, which seems to be reasonable.

For cropping, I propose to:
1. convert target box to mask.
2. ```input *  mask``` to apply masking.
3. crop out the non-zero area.
4. You can see the gradients output in the written doctest. The gradients are essentially equal everywhere in the targetted area, which seems to be reasonable.

The good thing of this API is as described in the first comment.

Jian

**Comment by edgarriba** [comment]
@shijianjian I think we already merged a patch to improve this

**Comment by edgarriba** [comment]
@shijianjian is this still relevant ?
